File: src/engine.py
from color import Color
from image import Image
from point import Point
from ray import Ray
from scene import Scene
from sphere import Sphere
from utils import st_time
from vector import Vector3


class RenderEngine:
    """Renders 3D object into 2D using raytracing"""
    MAX_DEPTH = 5
    MAX_DISTANCE = 10.0
    MIN_DISPLACE = 0.0001

    def __init__(self):
        self.specular_k = 50

    @st_time
    def render(self, scene: Scene) -> Image:
        """Render scene to image"""
        width, height = scene.width, scene.height
        aspect_ratio = width / height
        x0, x1 = -1.0, 1.0
        y0, y1 = x0 / aspect_ratio, x1 / aspect_ratio
        dx, dy = (x1 - x0) / (width - 1), (y1 - y0) / (height - 1)

        pixels = Image(width, height)

        for iy in range(height):
            y = y0 + iy * dy
            for ix in range(width):
                x = x0 + ix * dx
                ray = Ray(scene.camera, Point(x, y, 0) - scene.camera)
                pixels.set_pixel(ix, iy, self.ray_trace(ray, scene))
            # progress
            print(f"{iy/height:3.1%}", end="\r")
        return pixels

    def ray_trace(self, ray: Ray, scene: Scene, depth:int = 0) -> Color:
        """Cast ray and calculate color"""
        color = Color.BLACK
        distance_hit, object_hit = self.find_nearest(ray, scene)
        if object_hit is None or distance_hit > self.MAX_DISTANCE:
            return color
        hit_position = ray.origin + ray.direction * distance_hit
        hit_normal = object_hit.normal(hit_position)
        color += self.color_at(object_hit, hit_position, hit_normal, scene)

        # trace bounce ray
        if depth < self.MAX_DEPTH:
            new_ray = Ray(
                hit_position + hit_normal * self.MIN_DISPLACE,
                ray.direction - 2 * ray.direction.dot(hit_normal) * hit_normal
            )
            color += self.ray_trace(new_ray, scene, depth+1) * object_hit.material.reflection

        return color

    def find_nearest(self, ray: Ray, scene: Scene) -> Sphere:
        """Find the closest object hit by ray, return distance and object"""
        dist_min, objet_hit = None, None
        for obj in scene.objects:
            dist = obj.intersects(ray)
            if dist is not None and (objet_hit is None or dist < dist_min):
                dist_min, objet_hit = dist, obj
        return dist_min, objet_hit

    def color_at(self, object_hit: Sphere, hit_position: Point, hit_normal: Vector3, scene: Scene) -> Color:
        """Get color at hit point from each light, by combing ambient, diffuse and specular color"""
        material = object_hit.material
        obj_color = material.color_at(hit_position)
        to_cam = scene.camera - hit_position
        color = material.ambient * Color.WHITE

        for light in scene.lights:
            to_light = Ray(hit_position, light.position - hit_position)
            # Diffuse shading (Lambert)
            color += obj_color * material.diffuse * max(hit_normal.dot(to_light.direction), 0)
            # Specular shading (Blinn-Phong)
            half_vector = (to_light.direction + to_cam).normalize()
            color += light.color * material.specular * max(hit_normal.dot(half_vector), 0) ** self.specular_k
        return color

File: rust-tracer/bkp/engine.rs
use std::time::Instant;
use crate::{Color, Image, Point, Ray, Scene, Vector3, Sphere};

pub struct RenderEngine {
    max_depth: i32,
    max_distance: f64,
    min_displace: f64,
    specular_k: i32,
}

impl RenderEngine {
    pub fn new() -> Self {
        RenderEngine {
            max_depth: 5,
            max_distance: 10.0,
            min_displace: 0.0001,
            specular_k: 50,
        }
    }

    pub fn render(&self, scene: &Scene) -> Image {
        let start_time = Instant::now();
        
        let width = scene.width;
        let height = scene.height;
        let aspect_ratio = width as f64 / height as f64;
        let (x0, x1) = (-1.0, 1.0);
        let (y0, y1) = (x0 / aspect_ratio, x1 / aspect_ratio);
        let (dx, dy) = (
            (x1 - x0) / (width - 1) as f64,
            (y1 - y0) / (height - 1) as f64
        );

        let mut image = Image::new(width, height);

        for iy in 0..height {
            let y = y0 + iy as f64 * dy;
            for ix in 0..width {
                let x = x0 + ix as f64 * dx;
                let ray = Ray::new(
                    scene.camera,
                    Point::new(x, y, 0.0) - scene.camera
                );
                let color = self.ray_trace(&ray, scene, 0);
                image.set_pixel(ix, iy, color);
            }
            // Progress indicator
            print!("\rRendering: {:.1}%", (iy as f64 / height as f64) * 100.0);
        }

        let duration = start_time.elapsed();
        println!("\nRendering completed in {:.2?}", duration);
        
        image
    }

    fn ray_trace(&self, ray: &Ray, scene: &Scene, depth: i32) -> Color {
        let mut color = Color::BLACK;
        
        if let Some((distance, object)) = self.find_nearest(ray, scene) {
            if distance > self.max_distance {
                return color;
            }

            let hit_pos = ray.origin + ray.direction * distance;
            let hit_normal = object.normal(hit_pos);
            color = color + self.color_at(object, hit_pos, hit_normal, scene);

            // Handle reflections
            if depth < self.max_depth {
                let new_ray_dir = ray.direction - hit_normal * (2.0 * ray.direction.dot(&hit_normal));
                let new_ray = Ray::new(
                    hit_pos + hit_normal * self.min_displace,
                    new_ray_dir
                );
                color = color + self.ray_trace(&new_ray, scene, depth + 1) * object.material.reflection;
            }
        }

        color
    }

    fn find_nearest(&self, ray: &Ray, scene: &Scene) -> Option<(f64, &Sphere)> {
        let mut nearest: Option<(f64, &Sphere)> = None;

        for object in &scene.objects {
            if let Some(dist) = object.intersects(ray) {
                match nearest {
                    None => nearest = Some((dist, object)),
                    Some((min_dist, _)) if dist < min_dist => nearest = Some((dist, object)),
                    _ => {}
                }
            }
        }

        nearest
    }

    fn color_at(&self, object: &Sphere, hit_pos: Point, normal: Vector3, scene: &Scene) -> Color {
        let material = &object.material;
        let obj_color = material.color_at(hit_pos);
        let to_cam = scene.camera - hit_pos;
        let mut color = Color::WHITE * material.ambient;

        for light in &scene.lights {
            let to_light = Ray::new(
                hit_pos,
                light.position - hit_pos
            );

            // Diffuse shading (Lambert)
            let light_power = normal.dot(&to_light.direction).max(0.0);
            color = color + obj_color * material.diffuse * light_power;

            // Specular shading (Blinn-Phong)
            let half_vector = (to_light.direction + to_cam).normalize();
            let specular_power = normal.dot(&half_vector).max(0.0).powi(self.specular_k);
            color = color + light.color * material.specular * specular_power;
        }

        color
    }
}